### DISCUSSION TITLE
Bulk Upload Code Notes

### DISCUSSION THREAD

**Discussion by suXinjke** [RetroAchievements/RAWeb#1148] [Ideas]
The only user-friendly way to upload code notes right now is through RAIntegration, this is fine for small games and/or games that naturally don't require much notes to make achievements.

But look at people making sets for RPG games etc. - lots of code notes: hundreds, sometimes over thousand? It hurts to think of how much time they waste clicking the buttons one-by-one.

Now, technically it's possible to automate it by doing proper HTTP requests (which ain't easy to look at because it's RAIntegration doing them, not browser) and Javascript, but that's not user-friendly and would also require lots of responsibility - you must not _accidentally_ overwrite existing code notes incorrectly. It will also be slow because you still upload one-by-one and also have to respect rate limits.

So here's the proposal: make a separate HTTP endpoint that allows to bulk send code notes. The most important part is that it should only allow either to add new code notes, or update only the ones _you've uploaded_ - to prevent accidental overwrites of other people's notes.

I'm not sure what kind of format the backend would prefer, but I personally see regular JSON or CSV being the most convenient, like
```
// json
[
	["0xCAFECAFE", "Note [32-bit]\nCommentary"],
	["3405695742", "This note's address is decimal format of 0xCAFECAFE"]
]

// csv
"0xCAFECAFE"|"Note [32-bit]\nCommentary"
"3405695742"|"This note's address is decimal format of 0xCAFECAFE"
```

Now the mere existence of endpoint will be convenient enough - because other people can make simple tools or Javascript code to do the actual upload, those can also be shared. Ideally there should be a dedicated webpage with a form that accepts bulk data.

The data to upload would likely be generated from Google Sheets which achievement developers often use. I have an impression people are rather skilled with sheets to be able to generate all code notes rather easily.

**Comment by Chew** [comment]
This seems like it will partially be implemented by #4987 and https://github.com/RetroAchievements/rcheevos/pull/523
